sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/JoinOptimizationSuite.scala

```diff
@@ -123,6 +136,31 @@ class JoinOptimizationSuite extends PlanTest {
     }
   }
 
+  test("reorder inner joins - don't put predicate with IN subquery into join condition") {
+    // ReorderJoin collects all predicates and try to put them into join condition when creating
+    // ordered join. If a predicate with an IN subquery is in a join condition instead of a filter
+    // condition, `RewritePredicateSubquery.rewriteExistentialExpr` would fail to convert the
+    // subquery to an ExistenceJoin, and thus result in error.
+    val x = testRelation.subquery('x)
+    val y = testRelation1.subquery('y)
+    val z = testRelation.subquery('z)
+    val w = testRelation1.subquery('w)
+    val exists: AttributeReference = AttributeReference("exists", BooleanType, nullable = false)()
+
+    val queryPlan = x.join(y).join(z)
+      .where(("x.b".attr === "z.b".attr) && ("y.d".attr === "z.a".attr) &&
+        ("x.a".attr > 1 || "z.c".attr.in(ListQuery(w.select("w.d".attr)))))
+
+    val expectedPlan = x.join(z, Inner, Some("x.b".attr === "z.b".attr))
+      .join(w, ExistenceJoin(exists), Some("z.c".attr === "w.d".attr))
+      .where("x.a".attr > 1 || exists)
+      .select("x.a".attr, "x.b".attr, "x.c".attr, "z.a".attr, "z.b".attr, "z.c".attr)
+      .join(y, Inner, Some("y.d".attr === "z.a".attr))
```

Comment:
May I suggest the following:
- Change the title of the PR to "Preventing push down of IN subquery to Join operator"
- Update the test case to its bare minimum by removing the join to table y
- Change the title of the test case to reflect the new test

On the last point, removing the "reorder inner joins - " should do.